src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -621,3 +645,385 @@ def _find_basic_blocks(self, nodelist):
             basic_blocks.append(current_block)
             control_flow_nodes.append(None)
         return control_flow_nodes, basic_blocks
+
+    def _compute_backward_uses(self, basic_block_list):
+        """
+        Compute the backward uses for self._reference for the
+        basic_block_list provided. This function will not work
+        correctly if there is control flow inside the
+        basic_block_list provided.
+        The basic_block_list will be reversed to find the backward
+        accesses.
+        Reads to the reference that occur before a write will
+        be added to the self._uses array, the earliest write will
+        be provided as self._defsout and all previous writes
+        will be inside self._killed.
+
+        :param basic_block_list: The list of nodes that make up the basic
+                                 block to find the forward uses in.
+        :type basic_block_list: list[:py:class:`psyclone.psyir.nodes.Node`]
+
+        :raises NotImplementedError: If a GOTO statement is found in the code
+                                     region.
+        """
+        sig, _ = self._reference.get_signature_and_indices()
+        # For a basic block we will only ever have one defsout
+        defs_out = None
+        # Working backwards so reverse the basic_block_list
+        basic_block_list.reverse()
+        stop_position = self._stop_point
+        for region in basic_block_list:
+            region_list = region.walk((Reference, Call, CodeBlock, Return))
+            # If the region contains any Return, Exit or Cycle statements then
+            # we modify the stop position to only look at statements that
+            # occur before this statement.
+            # This doesn't work correctly if the Reference that
+            # is having its backwards dependencies analysed occurs after
+            # one of these such statements in a basic block, however
+            # since they're unreachable maybe we don't care?
```

Comment:
I agree, this is more evidence that we need a DeadCodeEliminationTrans to run early in the process so we get rid of all this.

Reply:
I guess is this a transformation we'd want to run in the frontend? Or is there a good reason to leave dead code?